controller/api/destination/endpoint_listener.go

```diff
@@ -204,46 +207,44 @@ func (l *endpointListener) toAddrSet(addresses []*updateAddress) *pb.AddrSet {
 }
 
 func (l *endpointListener) getAddrMetadata(pod *coreV1.Pod) (map[string]string, *pb.ProtocolHint, *pb.TlsIdentity) {
-	controllerNs := pod.Labels[pkgK8s.ControllerNSLabel]
+	controllerNS := pod.Labels[pkgK8s.ControllerNSLabel]
 	ownerKind, ownerName := l.ownerKindAndName(pod)
 	labels := pkgK8s.GetPodLabels(ownerKind, ownerName, pod)
 
-	var hint *pb.ProtocolHint
-
 	// If the pod is controlled by us, then it can be hinted that this destination
 	// knows H2 (and handles our orig-proto translation). Note that this check
 	// does not verify that the pod's control plane matches the control plane
 	// where the destination service is running; all pods injected for all control
 	// planes are considered valid for providing the H2 hint.
-	if l.enableH2Upgrade && controllerNs != "" {
+	var hint *pb.ProtocolHint
+	if l.enableH2Upgrade && controllerNS != "" {
 		hint = &pb.ProtocolHint{
 			Protocol: &pb.ProtocolHint_H2_{
 				H2: &pb.ProtocolHint_H2{},
 			},
 		}
 	}
 
-	if !l.enableTLS {
-		return labels, hint, nil
-	}
-
-	identity := pkgK8s.TLSIdentity{
-		Name:                ownerName,
-		Kind:                ownerKind,
-		Namespace:           pod.Namespace,
-		ControllerNamespace: controllerNs,
-	}
-
-	dnsName := identity.ToDNSName()
-
-	l.log.Debugf("getAddrMetadata(%+v): Owner: %s/%s DNS Name: %s Labels: %+v", pod, ownerKind, ownerName, dnsName, labels)
-
-	return labels, hint, &pb.TlsIdentity{
-		Strategy: &pb.TlsIdentity_K8SPodIdentity_{
-			K8SPodIdentity: &pb.TlsIdentity_K8SPodIdentity{
-				PodIdentity:  dnsName,
-				ControllerNs: controllerNs,
+	var identity *pb.TlsIdentity
+	if l.enableTLS && controllerNS == l.controllerNS &&
+		pod.Annotations[pkgK8s.IdentityModeAnnotation] == pkgK8s.IdentityModeOptional {
+		name := pkgK8s.TLSIdentity{
+			Name:                ownerName,
+			Kind:                ownerKind,
+			Namespace:           pod.Namespace,
+			ControllerNamespace: controllerNS,
+		}.ToDNSName()
+
+		l.log.Debugf("getAddrMetadata(%+v): Owner: %s/%s DNS Name: %s Labels: %+v", pod, ownerKind, ownerName, name, labels)
```

Comment:
(bump)

Reply:
I'm going to skip the controller namespace, since that will never change between log lines.

Reply:
Ok, I took a bigger whack at improving debug logging by consolidating the logic in endpointListener.Update:

Now, on updates, we log both additions and removals (previously, only additions were logged). Furthermore, we print the pod name and IP (before the IP wasn't represented as a dotted quad), and then include the entire protobuf weightedaddr struct (rather than a few select fields):

```
time="2019-02-26T16:53:48Z" level=debug msg="Update: add=1; remove=0" component=endpoint-listener namespace=linkerd service=linkerd-prometheus
time="2019-02-26T16:53:48Z" level=debug msg="Update: add: pod=linkerd-prometheus-7bbc899687-nd9zt; addr=10.1.1.160; addr:<ip:<ipv4:167838112 > port:9090 > weight:1 metric_labels:<key:\"control_plane_ns\" value:\"linkerd\" > metric_labels:<key:\"deployment\" value:\"linkerd-prometheus\" > metric_labels:<key:\"pod\" value:\"linkerd-prometheus-7bbc899687-nd9zt\" > metric_labels:<key:\"pod_template_hash\" value:\"7bbc899687\" > protocol_hint:<h2:<> > " component=endpoint-listener namespace=linkerd service=linkerd-prometheus
```